port/port_windows.cc

```diff
@@ -0,0 +1,149 @@
+// modification, are permitted provided that the following conditions are met:
+//
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of the University of California, Berkeley nor the
+//    names of its contributors may be used to endorse or promote products
+//    derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE REGENTS AND CONTRIBUTORS BE LIABLE FOR ANY
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+// ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include "port/port_windows.h"
+
+#include <windows.h>
+#include <cassert>
+
+namespace leveldb {
+namespace port {
+
+Mutex::Mutex() :
+    cs_(nullptr) {
+  assert(!cs_);
+  cs_ = static_cast<void *>(new CRITICAL_SECTION());
+  ::InitializeCriticalSection(static_cast<CRITICAL_SECTION *>(cs_));
+  assert(cs_);
+}
+
+Mutex::~Mutex() {
+  assert(cs_);
+  ::DeleteCriticalSection(static_cast<CRITICAL_SECTION *>(cs_));
+  delete static_cast<CRITICAL_SECTION *>(cs_);
+  cs_ = nullptr;
+  assert(!cs_);
+}
+
+void Mutex::Lock() {
+  assert(cs_);
+  ::EnterCriticalSection(static_cast<CRITICAL_SECTION *>(cs_));
+}
+
+void Mutex::Unlock() {
+  assert(cs_);
+  ::LeaveCriticalSection(static_cast<CRITICAL_SECTION *>(cs_));
+}
+
+void Mutex::AssertHeld() {
+  assert(cs_);
+  assert(1);
+}
+
+CondVar::CondVar(Mutex* mu) :
+    waiting_(0),
+    mu_(mu),
+    sem1_(::CreateSemaphore(NULL, 0, 10000, NULL)),
+    sem2_(::CreateSemaphore(NULL, 0, 10000, NULL)) {
+  assert(mu_);
+}
+
+CondVar::~CondVar() {
+  ::CloseHandle(sem1_);
+  ::CloseHandle(sem2_);
+}
+
+void CondVar::Wait() {
+  mu_->AssertHeld();
+
+  wait_mtx_.Lock();
+  ++waiting_;
+  wait_mtx_.Unlock();
+
+  mu_->Unlock();
+
+  // initiate handshake
+  ::WaitForSingleObject(sem1_, INFINITE);
+  ::ReleaseSemaphore(sem2_, 1, NULL);
+  mu_->Lock();
+}
+
+void CondVar::Signal() {
+  wait_mtx_.Lock();
+  if (waiting_ > 0) {
+    --waiting_;
+
+    // finalize handshake
+    ::ReleaseSemaphore(sem1_, 1, NULL);
+    ::WaitForSingleObject(sem2_, INFINITE);
```

Comment:
This is tricky code.  I would be much happier if we can use some existing library.  std::condition_variable seems the most appropriate to me.

Reply:
@ghemawat This has been removed. See the new code. I've used a centralized `port_cpp11` for this, and the new code is:

```cpp
std::unique_lock<std::mutex> lock(mu_->mu_, std::defer_lock);
cv_.wait(lock);
```

The rationale is the lock is already locked, and condition variable needs a unique_lock. Fortunately, `defer_lock` allows us to avoid unlocking and relocking the mutex. There is still an issue here though, in that the mutex is unlocked after waiting, which was not true previously (danger zone).

I am contemplating changing the above code to:

```cpp
{
    std::unique_lock<std::mutex> lock(mu_->mu_, std::defer_lock);
    cv_.wait(lock);
}
mu_->mu_->Lock();
```

This adds a redundant lock call, but reinstates correct behavior. Some comment would be great here.